lsp-types/src/Language/LSP/MetaModel/Types.hs

```diff
@@ -0,0 +1,225 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeApplications      #-}
+{-|
+This defines the types of the LSP "metamodel", which is a machine-readable format specifying the
+types used in the LSP protocol.
+
+The type system is quite typescript-y, which isn't surprising given that the whole protocol is
+very typescript-y.
+
+A typescript version of the metamodel types can be found here, which is useful for constructing
+this Haskell model of them:
+https://github.com/microsoft/vscode-languageserver-node/blob/main/tools/src/metaModel.ts
+-}
+module Language.LSP.MetaModel.Types where
+
+import           Data.Aeson         hiding (Null, String)
+import qualified Data.Aeson         as JSON
+import qualified Data.Aeson.TH      as JSON
```

Comment:
nit: some mixture between qualified and unqualified aeson functions in this file